Support new option "github-user"

yyoshiki41 opened the pull request:
## Description

Add new option `github-user`.

## Motivation and Context

That allowed the specified github users that is not on team or org.
e.g.) Username list (comma-separated)
"github-user:octocat,yyoshiki41"

fixed below issue.
https://github.com/pusher/oauth2_proxy/issues/342

## How Has This Been Tested?

My develop environment works this option.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pusher/oauth2_proxy/421)
<!-- Reviewable:end -->

tg44 commented:
I'm not a pro, nor a maintainer, but this seems to be ok.

JoelSpeed commented on main.go:
```diff
@@ -64,6 +64,7 @@ func main() {
 	flagSet.String("bitbucket-repository", "", "restrict logins to user with access to this repository")
 	flagSet.String("github-org", "", "restrict logins to members of this organisation")
 	flagSet.String("github-team", "", "restrict logins to members of this team")
+	flagSet.String("github-user", "", "restrict logins to these users")
```
We have a number of options that allow the flag to be specified multiple times, eg `whitelist-domain`, would it be possible to follow the same pattern?

yyoshiki41 replied:
fixed by using `flagSet.StringSlice`.

yyoshiki41 commented:
@JoelSpeed 

Sorry for late reply.
I fixed some issues and pushed.
- [ ] rebase master
- [ ] fix tests
- [ ] add some tests
- [ ] update docs

Thanks!

JoelSpeed commented on options.go:
```diff
@@ -56,6 +56,7 @@ type Options struct {
 	GitHubTeam               string   `flag:"github-team" cfg:"github_team" env:"OAUTH2_PROXY_GITHUB_TEAM"`
 	GitHubRepo               string   `flag:"github-repo" cfg:"github_repo" env:"OAUTH2_PROXY_GITHUB_REPO"`
 	GitHubToken              string   `flag:"github-token" cfg:"github_token" env:"OAUTH2_PROXY_GITHUB_TOKEN"`
+	GitHubUsers              []string `flag:"github-user" cfg:"github_user" env:"OAUTH2_PROXY_GITHUB_USER"`
```
To be consistent with other list options, will also need to update the docs
```suggestion
	GitHubUsers              []string `flag:"github-user" cfg:"github_users" env:"OAUTH2_PROXY_GITHUB_USERS"`
```

yyoshiki41 replied:
https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/8952cfb01c99661f478465790477db58e29720b1
and updated docs: https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/be7f969902b11d75b0af877a153815d4fc912fdf.

JoelSpeed commented on providers/github.go:
```diff
@@ -80,6 +83,19 @@ func (p *GitHubProvider) SetRepo(repo, token string) {
 
+// SetUsers configures allowed usernames
+func (p *GitHubProvider) SetUsers(users []string) {
+	for _, user := range users {
+		us := strings.Split(user, ",")
+		for _, u := range us {
+			// Validate empty string (e.g. --github-user="octcat,,")
+			if len(u) > 0 {
+				p.Users = append(p.Users, u)
+			}
+		}
+	}
```
This will be covered by viper, the configuration loader

JoelSpeed commented on providers/github.go:
```diff
@@ -317,6 +333,48 @@ func (p *GitHubProvider) hasRepo(ctx context.Context, accessToken string) (bool,
+	}
+
+	if p.isVerifiedUser(user.Login) {
+		log.Printf("Found Github User: %q", user.Login)
```
Will this log line become noisy? Same for the one below?

JoelSpeed commented on providers/github.go:
```diff
@@ -356,21 +414,38 @@ func (p *GitHubProvider) GetEmailAddress(ctx context.Context, s *sessions.Sessio
-		} else {
-			if ok, err := p.hasOrg(ctx, s.AccessToken); err != nil || !ok {
-				return "", err
+		} else if p.Repo != "" {
+			if p.Token == "" { // If we have a token we'll do the collaborator check in GetUserName
```
Why split these two up?

yyoshiki41 commented:
@JoelSpeed 
Thanks for your review!
I fixed all your review points and pushed.

JoelSpeed commented on options.go:
```diff
@@ -56,6 +56,7 @@ type Options struct {
 	GitHubTeam               string   `flag:"github-team" cfg:"github_team" env:"OAUTH2_PROXY_GITHUB_TEAM"`
 	GitHubRepo               string   `flag:"github-repo" cfg:"github_repo" env:"OAUTH2_PROXY_GITHUB_REPO"`
 	GitHubToken              string   `flag:"github-token" cfg:"github_token" env:"OAUTH2_PROXY_GITHUB_TOKEN"`
+	GitHubUsers              []string `flag:"github-users" cfg:"github_users" env:"OAUTH2_PROXY_GITHUB_USERS"`
```
Sorry, I wasn't very clear in my previous comment. The flag should be singular, while the other two should be pluralised (yes this is very confusing, but it's what we have for now).
```suggestion
	GitHubUsers              []string `flag:"github-user" cfg:"github_users" env:"OAUTH2_PROXY_GITHUB_USERS"`
```

yyoshiki41 replied:
Thanks for your explanation, I understood!
https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/5dc11924b63c590f3b61f79ff0bac3a15107a138

JoelSpeed commented on providers/github.go:
```diff
@@ -317,6 +325,47 @@ func (p *GitHubProvider) hasRepo(ctx context.Context, accessToken string) (bool,
+	if p.isVerifiedUser(user.Login) {
+		return true, nil
+	}
+	log.Printf("Missing Github User: %q in %v", user.Login, p.Users)
```
I think we should drop this log too, could get noisy

yyoshiki41 replied:
Remove logging https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/19f9b34ce854faf0e296cc9c87ce4270649c207d

JoelSpeed commented:
@yyoshiki41 Could you please rebase this and then I'll give it another review

yyoshiki41 commented:
@JoelSpeed 
rebased and move options to follow bellow pr
https://github.com/oauth2-proxy/oauth2-proxy/pull/489

Thanks!

JoelSpeed commented on providers/github_test.go:
```diff
@@ -318,3 +318,78 @@ func TestGitHubProviderGetUserNameWithRepoAndTokenWithoutPushAccess(t *testing.T
+
+	bURL, _ := url.Parse(b.URL)
+	p := testGitHubProvider(bURL.Host)
+	p.SetUsers([]string{"mbland", "octcat"})
```
```suggestion
	p.SetUsers([]string{"mbland", "octocat"})
```

JoelSpeed commented on providers/github_test.go:
```diff
@@ -318,3 +318,78 @@ func TestGitHubProviderGetUserNameWithRepoAndTokenWithoutPushAccess(t *testing.T
+
+	bURL, _ := url.Parse(b.URL)
+	p := testGitHubProvider(bURL.Host)
+	p.SetUsers([]string{"octcat"})
```
```suggestion
	p.SetUsers([]string{"octocat"})
```

yyoshiki41 commented:
@JoelSpeed 

fixed to improve readability in the docs
https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/0333f9f414c96fb9c34725fd10d506fcf4416e65

sorry, I missed your review.
I fixed it by https://github.com/oauth2-proxy/oauth2-proxy/pull/421/commits/86cadbfbc6b4edcdb1981e0a561188c86caf3f5b

I fixed all your review points .
Thanks.

yyoshiki41 commented on providers/github.go:
```diff
@@ -80,6 +82,11 @@ func (p *GitHubProvider) SetRepo(repo, token string) {
 
+// SetUsers configures allowed usernames
+func (p *GitHubProvider) SetUsers(users []string) {
+	p.Users = append(p.Users, users...)
```
Sorry, I missed this points.
fixed by 86cadbf

JoelSpeed commented on pkg/apis/options/options.go:
```diff
@@ -228,6 +229,7 @@ func NewFlagSet() *pflag.FlagSet {
 	flagSet.String("github-team", "", "restrict logins to members of this team")
 	flagSet.String("github-repo", "", "restrict logins to collaborators of this repository")
 	flagSet.String("github-token", "", "the token to use when verifying repository collaborators (must have push access to the repository)")
+	flagSet.StringSlice("github-user", []string{}, "allowed users with these usernames to login even if they do not belong to the specified org and team or collaborators (may be given multiple times)")
```
```suggestion
	flagSet.StringSlice("github-user", []string{}, "allow users with these usernames to login even if they do not belong to the specified org and team or collaborators (may be given multiple times)")
```
